packages/scanner-global-library/src/page-handler.ts

```diff
@@ -8,14 +8,50 @@ import { System } from 'common';
 
 @injectable()
 export class PageHandler {
-    constructor(@inject(GlobalLogger) @optional() private readonly logger: GlobalLogger) {}
+    constructor(
+        @inject(GlobalLogger) @optional() private readonly logger: GlobalLogger,
+        private readonly checkIntervalMsecs: number = 200,
+    ) {}
 
     public async waitForPageToCompleteRendering(
         page: Puppeteer.Page,
-        timeoutMsecs: number,
-        checkIntervalMsecs: number = 200,
+        scrollTimeoutMsecs: number,
+        renderTimeoutMsecs: number,
     ): Promise<void> {
-        const maxCheckCount = timeoutMsecs / checkIntervalMsecs;
+        // Scroll to the bottom of the page to resolve pending page operations and load lazily-rendered content
+        await this.scrollToBottom(page, scrollTimeoutMsecs);
+        await this.waitForStableContent(page, renderTimeoutMsecs);
+    }
+
+    private async scrollToBottom(page: Puppeteer.Page, timeoutMsecs: number): Promise<void> {
+        const maxCheckCount = timeoutMsecs / this.checkIntervalMsecs;
+        let checkCount = 0;
+        let scrollingComplete = false;
+
+        // Scroll incrementally so everything is inside the window at some point
+        while (!scrollingComplete && checkCount < maxCheckCount) {
+            // Page evaluation may fail because of a navigation
+            try {
+                scrollingComplete = await page.evaluate(async () => {
+                    window.scrollBy(0, window.innerHeight);
+
+                    return window.document.scrollingElement.scrollTop + window.innerHeight >= window.document.scrollingElement.scrollHeight;
```

Comment:
do we need `return`?

Reply:
Yes, this is returning a boolean value for whether or not we've reached the end of the page